### ISSUE TITLE
Change Kybra to remove duplicate matches

### ISSUE THREAD

**Issue by bdemann** [demergent-labs/kybra#373]
Dan had an idea that in some cases we might be able to pass around the relevant insides for things so we don't have to go get the insides again when we are working on them. For example for to_tuple we need to start at the ExprKind Tuple[str] and dive in, make sure it's Tuple and then process str. Dan says why not just process str.

So how that would work is in to_candid_type instead of having is_prim or is_tuple etc we would create a get_type function that would look something like this

``` rust
impl SourceMapped<&Located<ExprKind>> {
    pub fn get_type(&self) -> ExprKindType {
        if self.is_primitive() {
            let primitive = match &self.node {
                ExprKind::Name { id, .. } => ExprKindType::Primitive(id.to_string()),
                _ => panic!(),
            };
            return primitive;
        }
        if self.is_tuple() {
            return match &self.node {
                ExprKind::Subscript { slice, .. } => {
                    ExprKindType::Tuple(SourceMapped::new(slice, self.source_map.clone()))
                }
                _ => panic!(),
            };
        }
        panic!()
    }
}
```

And then in to_candid type we would do 
```rust
        return match self.get_type() {
            ExprKindType::Primitive(_) => todo!(),
            ExprKindType::Array => todo!(),
            ExprKindType::Opt => todo!(),
            ExprKindType::Tuple(slice) => slice.to_new_tuple(None),
            ExprKindType::TypeRef => todo!(),
        };
```

And then new_tuple would look like:
```rust
pub fn to_new_tuple(&self, tuple_name: Option<String>) -> Result<Tuple, Vec<Error>> {
        let tuple_members_exprs = match &self.node {
            ExprKind::Tuple { elts, .. } => elts
                .iter()
                .map(|elt| SourceMapped::new(elt, self.source_map.clone()))
                .collect(),
            _ => {
                vec![SourceMapped::new(
                    self.deref().clone(),
                    self.source_map.clone(),
                )]
            }
        };
        let elems = tuple_members_exprs
            .iter()
            .map(|kybra_elem| kybra_elem.as_tuple_member())
            .collect_results()?;
        Ok(Tuple {
            name: tuple_name,
            elems,
            type_params: vec![].into(),
        })
    }
```
As opposed to the old way
```rust
pub fn to_tuple(&self, tuple_name: Option<String>) -> Result<Tuple, Vec<Error>> {
        if !self.is_tuple() {
            return Err(self.not_tuple_error().into());
        }
        match &self.node {
            ExprKind::Subscript { slice, .. } => {
                let tuple_members_exprs = match &slice.node {
                    ExprKind::Tuple { elts, .. } => elts
                        .iter()
                        .map(|elt| SourceMapped::new(elt, self.source_map.clone()))
                        .collect(),
                    _ => {
                        vec![SourceMapped::new(slice.as_ref(), self.source_map.clone())]
                    }
                };
                let elems = tuple_members_exprs
                    .iter()
                    .map(|kybra_elem| kybra_elem.as_tuple_member())
                    .collect_results()?;
                Ok(Tuple {
                    name: tuple_name,
                    elems,
                    type_params: vec![].into(),
                })
            }
            _ => Err(Unreachable::new_err().into()),
        }
    }
```
which as you can see still has to check if it's a tuple, and dive through the subscript layer. 
It should be noted that not only does the to_new_tuple() not need to check if it's a tuple but it does not have the ability to do so. Which one could argue get's rid of the issue of "what kind of error should we return if this function is called inappropriately" and replaces it with not being able to know if the function was called inappropriately. But it does remove a bit of the extra peeling back the layers. 

Long story short, we might consider doing this, but maybe we don't do it. For right now I am holding off because it seems like a pretty big change.

**Comment by lastmjs** [comment]
We are closing this issue anticipating an entire rearchitecture of Kybra similar to what we have done with Azle. This removes most of the compiler architecture as it is. We believe that this issue no longer applies to the new planned architecture.

In case our rearchitecture is not successful, or we otherwise need to find these closed issues, the search term KYBRA_REARCHITECTURE can be used.
